src/main/groovy/com/cloudcard/photoDownloader/OrigoClient.groovy

```diff
@@ -0,0 +1,211 @@
+package com.cloudcard.photoDownloader
+
+import jakarta.annotation.PostConstruct
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.stereotype.Component
+import com.fasterxml.jackson.databind.ObjectMapper
+
+import static com.cloudcard.photoDownloader.ApplicationPropertiesValidator.throwIfBlank
+
+@Component
+@ConditionalOnProperty(value = "downloader.storageService", havingValue = "OrigoStorageService")
+class OrigoClient {
+    // Makes requests to Origo API
+
+    private static final Logger log = LoggerFactory.getLogger(OrigoClient.class)
+
+    @Autowired
+    HttpClient httpClient
+
+    @Autowired
+    UnirestWrapper unirestWrapper
+
+    @Value('${Origo.eventManagementApi}')
+    String eventManagementApi
+
+    @Value('${Origo.mobileIdentitiesApi}')
+    String mobileIdentitiesApi
+
+    @Value('${Origo.certIdpApi}')
+    String certIdpApi
+
+    @Value('${Origo.organizationId}')
+    String organizationId
+
+    @Value('${Origo.contentType}')
+    String contentType
+
+    @Value('${Origo.applicationVersion}')
+    String applicationVersion
+
+    @Value('${Origo.applicationId}')
+    String applicationId
+
+    @Value('${Origo.clientId}')
+    String clientId
+
+    @Value('${Origo.clientSecret}')
+    String clientSecret
+
+    String accessToken
+
+    boolean isAuthenticated = false
+
+    Map requestHeaders
+
+    @PostConstruct
+    init() {
+        throwIfBlank(eventManagementApi, "The Origo Event Management API URL must be specified.")
+        throwIfBlank(mobileIdentitiesApi, "The Origo Mobile Identities API URL must be specified.")
+        throwIfBlank(organizationId, "The Origo organization ID must be specified.")
+        throwIfBlank(clientSecret, "The Origo client secret must be specified.")
+        throwIfBlank(clientId, "The Origo client ID must be specified.")
+        throwIfBlank(contentType, "The Origo content-type header must be specified.")
+        throwIfBlank(applicationVersion, "The Origo application version must be specified.")
+        throwIfBlank(applicationId, "Your organization's Origo Application ID must be specified.")
+
+        log.info('=================== Initializing Origo Client ===================')
+        log.info("          Origo Event Management API URL : $eventManagementApi")
+        log.info("         Origo Mobile Identities API URL : $mobileIdentitiesApi")
+        log.info("                   Origo organization ID : $organizationId")
+        log.info("               Origo content type header : $contentType")
+        log.info("               Origo application version : $applicationVersion")
+        log.info("                    Origo application ID : $applicationId")
+
+        httpClient.source = this.class.name
+
+    }
+
```

Comment:
I would consider refactoring this one a bit. Returning a boolean for a method that does an action (authenticate) is a little confusing. We can probably make it a void method that just toggles the isAuthenticated boolean.

Reply:
Also, it looks like tokens expire after an hour. Do we have a way to request a new token on expiry?

Reply:
Refactored into a void method. Yes, the next commit that I publish will pass all requests through a "makeAuthenticatedRequest" method that handles re-authentication automatically.